feat(config): make `output.library` accept functions

byzyk opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. -->
<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->


<!-- In addition to that please answer these questions: -->

**What kind of change does this PR introduce?**

feature, closes #7904 

**Did you add tests for your changes?**

Yes

**Does this PR introduce a breaking change?**

No

**What needs to be documented once your changes are merged?**

`output.library`, `output.libraryExport` can now accept function. 
<!-- List all the information that needs to be added to the documentation after merge -->
<!-- When your changes are merged you will be asked to contribute this to the documentation -->

evilebottnawi commented:
@byzyk thanks for PR! What is use case?

byzyk commented:
@evilebottnawi feature was requested by a user as "nice to have", please have a look at #7904 for the use case.

byzyk commented on lib/LibraryTemplatePlugin.js:
```diff
@@ -40,15 +40,15 @@ const accessorAccess = (base, accessor, joinWith = "; ") => {
 	constructor(name, target, umdNamedDefine, auxiliaryComment, exportProperty) {
-		this.name = name;
-		this.target = target;
+		this.name = typeof name === "function" ? name() : name;
```
thanks for the feedback @sokra.

Do you think `WebpackOptionsApply.js` is a good place to evaluate the function?

byzyk replied:
> Pass an object to the function `{ chunk }`

Correct me if I'm wrong but isn't it already done in [`TemplatePathPlugin`](https://github.com/webpack/webpack/blob/master/lib/TemplatedPathPlugin.js#L76)?

byzyk replied:
Ok, I'll take another look into that. What do you think would be an appropriate place to evaluate the functions?

sokra replied:
Pass the function from LibraryTemplatePlugin into the inner plugins `SetVarMainTemplatePlugin`, `AmdMainTemplatePlugin`, `UmdMainTemplatePlugin` and `JsonpExportMainTemplatePlugin`. Make sure these plugins accept functions.

As it's allowed to pass a function to `getAssetPath` you should be good to go in most cases.

byzyk commented on lib/LibraryTemplatePlugin.js:
```diff
@@ -23,7 +23,14 @@ const accessorToObjectAccess = accessor => {
+		accessors = [accessor()];
+	} else {
+		accessors = [accessor];
+	}
```
hey @sokra, I attempted to allow functions for lib names first. In the case of `SetVarMainTemplatePlugin` it's not possible to evaluate the function inside the plugin because in most of the cases this function goes as a first parameter, modifying `name` property provided by `LibraryTemplatePlugin`.

Do you feel like we should move this functionality inside `SetVarMainTemplatePlugin` since it's not used in other plugins?

sokra replied:
You don't have to change this function:

Instead of

``` js
 					new SetVarMainTemplatePlugin(
 						`var ${accessorAccess(undefined, this.name)}`,
 						false
 					).apply(compilation);
```


``` js
 					new SetVarMainTemplatePlugin(
 						typeof this.name === "function" ?
 							data => `var ${accessorAccess(undefined, this.name(data))}` :
 							`var ${accessorAccess(undefined, this.name)}`,
 						false
 					).apply(compilation);
```

byzyk replied:
Ok, I see your point. But isn't it gonna produce a lot of duplicate code? Perhaps I'm missing something but I don't understand why it has to be done this way.

sokra replied:
Instead of duplicating code you could move this code into a method:

``` js
_withName(fn) {
  return typeof this.name === "function" ?
    data => fn(this.name(data)) :
    fn(this.name)
}

new SetVarMainTemplatePlugin(
  this._withName(name => `var ${accessorAccess(undefined, name)}`)
  false
).apply(compilation);

byzyk commented:
Hi @sokra, I followed the approach you've suggested. Please let me know if that's something we're looking for?

Also, I think like those helper functions `_withName`, `_withTarget` can be moved to the top of `.apply()` since they are used quite extensively there.

sokra commented on lib/LibraryTemplatePlugin.js:
```diff
@@ -125,29 +129,48 @@ class LibraryTemplatePlugin {
 	}
+
+	_withName(fn) {
+		return typeof this.name === "function" ? fn(this.name()) : fn(this.name);
```
Please read my comment again. This is supposed to return a function when `this.name` is a function.

sokra commented on lib/LibraryTemplatePlugin.js:
```diff
@@ -125,29 +129,48 @@ class LibraryTemplatePlugin {
+		return typeof this.name === "function" ? fn(this.name()) : fn(this.name);
+	}
+
+	_withTarget(fn) {
```
Please revert all changes regarding `target`. The current approach doesn't work for `target` as function, since we can't apply different plugins per chunk.

sokra replied:
Concentrate on allow functions for `name`.

sokra commented on test/configCases/library/function-umd/webpack.config.js:
```diff
@@ -0,0 +1,11 @@
+module.exports = {
+	output: {
+		libraryTarget: () => "umd",
+		library: () => ({
```
The test case should feature multiple entrypoints which have different library values per name.

i. e. `library: ({ chunk }) => chunk.name`

byzyk commented:
Hey @sokra, I've reverted everything `target` related and added a couple of more tests to be more secure.

However, I'm still struggling to understand the appropriate place to eval the functions. Take a look at the failing tests, please.

`SetVarMainTemplatePlugin` seems to be working fine.

`UmdMainTemplatePlugin` the problem is that we also allow object as a valid input for library name here and it [gets checked inside the constructor](https://github.com/webpack/webpack/blob/master/lib/UmdMainTemplatePlugin.js#L64). I'm not sure if evaluating name function inside a constructor is doable in this case, what are your thoughts? I feel like object related stuff has to be moved out of constructor in this case.

`AmdMainTemplatePlugin` throws type error in `hash.update` function, although I'm not sure whether it's a good idea to invoke the name function there as well. I think like we should do that in `apply` method. Do you think like that's ok and would you mind guiding me a bit on how to get chunk data inside `apply` method so we could pass the data to the function?

`JsonpExportMainTemplatePlugin` not tested yet.

byzyk commented on lib/LibraryTemplatePlugin.js:
```diff
@@ -125,13 +127,15 @@ class LibraryTemplatePlugin {
 				case "umd2": {
 					const UmdMainTemplatePlugin = require("./UmdMainTemplatePlugin");
-					new UmdMainTemplatePlugin(this.name, {
+					new UmdMainTemplatePlugin(this._withName(name => name), {
```
I see, so we might only wanna leave this for `SetVarMainTemplatePlugin`, right?

sanci-me commented:
Thanks so much for what you have done already.  Is there any more progress on this feature?

vankop commented:
Thanks for your contribution. I am closing PR because it is too stale..
